kernel/sched.c

```diff
@@ -1265,39 +1265,44 @@ static void halt_thread(struct k_thread *thread, uint8_t new_state)
 	/* We hold the lock, and the thread is known not to be running
 	 * anywhere.
 	 */
-	if ((thread->base.thread_state & new_state) == 0U) {
+	if (likely((thread->base.thread_state & new_state) == 0U)) {
 		thread->base.thread_state |= new_state;
 		if (z_is_thread_queued(thread)) {
 			dequeue_thread(thread);
 		}
 
-		if (new_state == _THREAD_DEAD) {
-			if (thread->base.pended_on != NULL) {
-				unpend_thread_no_timeout(thread);
-			}
-			(void)z_abort_thread_timeout(thread);
-			unpend_all(&thread->join_queue);
+		if (likely(new_state == _THREAD_SUSPENDED)) {
```

Comment:
How impactful is this patch to halt_thread()?  This code was excruciatingly hard to get right and coming at it with a refactoring hatchet just for a minor performance boost gives me the willies.  I don't see anything wrong, but...

I guess my strong preference would be to split this one patch out into a separate PR and give it a ton of testing in isolation.

Reply:
This one patch gave us a 9.7% performance boost in the thread_metric preemptive benchmark (multiq) on the disco_l475_iot1 board when compared to the numbers from the previous commit.